### ISSUE TITLE
OTP-24.0 is miscompiled by gcc's LTO

### ISSUE THREAD

**Issue by mikpe** [erlang/otp#4846]
**Describe the bug**
When `gcc`'s link-time optimizations are enabled, via `-flto`, OTP-24.0 and current master are miscompiled, which shows itself as
an internal error in beam_kernel_to_ssa as the new VM invokes its `erlc` for the first time:

> make[1]: Leaving directory '/tmp/otp.git/erts'
> cd lib && \
>   ERL_TOP=/tmp/otp.git PATH=/tmp/otp.git/bootstrap/bin:"${PATH}" \
>         make opt SECONDARY_BOOTSTRAP=true ERL_COMPILE_WARNINGS_AS_ERRORS=yes
> make[1]: Entering directory '/tmp/otp.git/lib'
> make[2]: Entering directory '/tmp/otp.git/lib/parsetools'
> === Entering application parsetools
> make[3]: Entering directory '/tmp/otp.git/lib/parsetools/src'
> erlc -W  -Werror +debug_info -DUSE_ESOCK=true -I/tmp/otp.git/lib/stdlib/include -Werror -o../ebin leex.erl
> Function: compile/3
> leex.erl: internal error in pass beam_kernel_to_ssa:
> exception error: bad key: 1
>   in function  map_get/2
> ...

**To Reproduce**
Build current maint-24 or master with `CFLAGS="-O2 -flto"` and gcc-10.3. I see the error with `gcc version 10.3.1 20210422 (Red Hat 10.3.1-1) (GCC)` on Fedora 32 and Fedora 33. I have not tested other GCC versions yet, but it is persistent with this one.

**Expected behavior**
I expected the newly built VM to manage to run its compiler to build its Erlang modules.

**Affected versions**
OTP-24.0 and current master are affected. Current versions of OTP-23 and OTP-22 are _not_ affected.

**Additional context**
We build Erlang/OTP as an RPM, and `-flto` is added by default from `/usr/lib/rpm/redhat/macros`. We can work around the error by adding `%define _lto_cflags %{nil}` to our RPM `.spec` file, presumably trading a performance loss for correctness.

The build with `-flto` shows a scary warning:

> beam/erl_driver.h:331:12: warning: type of 'driver_select' does not match original declaration [-Wlto-type-mismatch]
  331 | EXTERN int driver_select(ErlDrvPort port, ErlDrvEvent event, int mode, int on);
      |            ^
sys/common/erl_check_io.c:796:1: note: type mismatch in parameter 2
  796 | driver_select(ErlDrvPort ix, ErlDrvEvent e, int mode, int on)
      | ^
sys/common/erl_check_io.c:796:1: note: type 'ErlDrvEvent' should match type 'struct _erl_drv_event *'
sys/common/erl_check_io.c:796:1: note: 'driver_select' was previously declared here
sys/common/erl_check_io.c:796:1: note: code may be misoptimized unless '-fno-strict-aliasing' is used

but this occurs also for OTP-23 where things work, so it may be a red herring. Still, the code seems internally inconsistent about what type `ErlDrvEvent` is: some places treat it as a struct pointer while others treat it as a signed integer.

**Comment by mikpe** [comment]
A git bisect between OTP-23.3.4 (good) and master identified

> # first bad commit: [efd2f8ea4b077320cd2fbe4b2a82d867548264bb] Optimize erts_qsort

Reverting that and two followups that modified the same piece of code, `d443f060872dc0484d7bc1d18d9834a312c104ea` and `124c211af8ee64a34035d556bccec68dc27b7b7d`, gives me a master that builds Ok with `-flto`.

**Comment by sverker** [comment]
FYI @mikpe: This is a strict-aliasing related bug causing our own erts_qsort() to produce incorrect result. I'm trying to figure out if this is our fault or if we can blame gcc-10.3.

**Comment by mikpe** [comment]
> FYI @mikpe: This is a strict-aliasing related bug causing our own erts_qsort() to produce incorrect result. I'm trying to figure out if this is our fault or if we can blame gcc-10.3.

Interesting. Do you have a stand-alone test case? If so I could try it out on different gcc versions and perhaps do a git bisect on gcc if there appears to be a regression in 10.3.

**Comment by sverker** [comment]
Here is a stand-alone test case that provokes the bug.
https://github.com/sverker/lto_bug

**Comment by frej** [comment]
Gcc 10.2.0 also triggers the `ASSERT: NOT HIGHER THAN PIVOT`, compiling with `-fno-strict-aliasing` does not trigger any assert. With Clang 12.0.0, the test-case executes successfully both with `-fstrict-aliasing`.

The type-punning in `erts_qsort_swap` is UB as we are accessing an object `iptr`/`jptr` which are `char*` through a `unsigned long*` which is neither "type compatible with the effective type", "a qualified version of a type compatible with the effective type of the object", "a type that is the signed or unsigned type corresponding to the effective type of the object", "a type that is the signed or unsigned type corresponding to a qualified version of the effective type of the object", "an aggregate or union type that includes one of the aforementioned types among its members (including, recursively, a member of a subaggregate or contained union)" nor "a character type".

Rewriting `erts_qsort_swap` into the rather ugly looking (assuming that item_size is at most a machine word):
```
static void erts_qsort_swap(unsigned item_size,
                            char* iptr,
                            char* jptr)
{
    char tmp[item_size];
    memcpy(tmp, jptr, item_size);
    memcpy(jptr, iptr, item_size);
    memcpy(iptr, tmp, item_size);
}

```
Gcc is smart enough to just emit a series of moves, although in this case, it doesn't use an xchg as it appears to reuse both values. If `item_size` is big, wrap the memcpys in a loop, but for this case the emitted code will be the same as `item_size` is known:
```
static void erts_qsort_swap(unsigned item_size,
                            char* iptr,
                            char* jptr)
{
  unsigned long tmp;
  unsigned cnt;

  for (cnt = item_size / sizeof(tmp); cnt; cnt--) {
    memcpy(&tmp, jptr, sizeof(tmp));
    memcpy(jptr, iptr, sizeof(tmp));
    memcpy(iptr, &tmp, sizeof(tmp));
    iptr += sizeof(tmp);
    jptr += sizeof(tmp);
  }
}

**Comment by sverker** [comment]
Thanks for the analysis, @frej.
So the char-pointer exception for strict aliasing does not help us here. Is it only when _dereferencing_ a char-pointer that strict aliasing is not applied towards other pointers? Casting via a char-pointer to a non compatible type does nothing wrt strict aliasing then?

The item_size is 24 (3 words) in the real use case.

**Comment by mikpe** [comment]
Also `hxnodecmp()` in the test case is wrong: its formal parameters _must_ be `const char *` since that's what `erts_sqort()` sees. So it needs to be:
```
static int hxnodecmp(const char *a0, const char *b0)
{
    const hxnode_t *a = a0, *b = b0;
    ...
```
The assignment compatibility between void*/char* and other (data) pointer types only applies when the compiler sees the different types and can perform any implicit conversion. When calling a function pointer the types must match _exactly_.

And as frej wrote, in the test case you use `long` to access `int` objects, which is UB.

glibc's `qsort()` does the swap byte-wise in a loop. I do think we want to unroll that, but to avoid UB use `memcpy()` via a `long`-sized `tmp`. (VLAs are bad so we don't want that alternative.)

**Comment by frej** [comment]
> Is it only when dereferencing a char-pointer that strict aliasing is not applied towards other pointers?

Yes, although the pointer may point to an object of an incompatible type, the behavior is defined and the compiler must assume that other pointers of a different type can alias (unless it can prove otherwise).

> Casting via a char-pointer to a non compatible type does nothing wrt strict aliasing then?

The standard only talks about "accessing an object", my understanding is that the cast is valid, but as soon as you dereference the pointer you enter undefined territory.

**Comment by sverker** [comment]
Yes, `hxnodecmp()` args was not correct, but fixing it did not do any difference for this problem.

glibc's `qsort()` does not assume any alignment at all, so that's why it must do byte wise swap I suppose.

**Comment by mikpe** [comment]
I created a PR at https://github.com/sverker/lto_bug with fixes that work for me with gcc-10.3 and gcc-11.1.

**Comment by sverker** [comment]
Yes, it works for me also with gcc-10.3.

I did some quick benchmarking and I can't see any difference with a memcpy swap.

I will create a PR for the OTP repo.

**Comment by sverker** [comment]
#4894 merged to maint for OTP-24.1. Thanks, @mikpe !
